Sources/LanguageServerProtocolTransport/MessageSplitting.swift

```diff
@@ -10,14 +10,117 @@
 //
 //===----------------------------------------------------------------------===//
 
-import Foundation
+package import Foundation
 import LanguageServerProtocol
+@_spi(SourceKitLSP) import SKLogging
 
+/// A stateful JSONRPC message parser. Expects data to be read in chunks of `nextReadLength`.
+package class JSONMessageParser<MessageType> {
+  /// Decoder of the JSON message contents
+  private let decoder: (Data) -> MessageType?
+
+  /// Buffer of received bytes that haven't been fully parsed.
+  private nonisolated(unsafe) var requestBuffer: [UInt8] = []
+
+  /// Current parser state
+  private var state: ReadState = .header
+
+  /// The number of bytes to read given the current state of the parser.
+  package var nextReadLength: Int {
+    switch state {
+    case .header:
+      // Only ever read a single byte for the header to better handle invalid cases
+      return 1
+    case .message(let remaining):
+      // Up until the message, where we should read its entire length (or anything remaining if we had a partial read)
+      return remaining
+    }
+  }
+
+  package init(decoder: @escaping (Data) -> MessageType?) {
+    self.decoder = decoder
+  }
+
+  /// Parse the next chunk of data (see `nextReadLength`). Returns a message parsed by `decoder` if read and `nil`
+  /// otherwise (including on error). Note that this does not handle being passed a chunk that crosses header+contents
+  /// boundaries or whole message boundaries.
+  package func parse(chunk: Data) -> MessageType? {
+    switch state {
+    case .header:
+      parseHeader(data: chunk)
+      return nil
+    case .message(let remaining):
+      return parseMessage(data: chunk, remaining: remaining)
+    }
+  }
+
+  func parseHeader(data: Data) {
+    requestBuffer.append(contentsOf: data)
+    if requestBuffer.suffix(4) != JSONRPCMessageHeader.headerSeparator {
+      return
+    }
+
+    let header = orLog("Parsing JSONRPC message") {
+      try requestBuffer.jsonrpcParseHeader()
+    }
+    requestBuffer.removeAll(keepingCapacity: true)
+
+    guard let header,
+      let length = header.contentLength
+    else {
+      logger.error("Ignoring message due to invalid header")
+      return
+    }
+
+    state = .message(remaining: length)
+  }
+
+  func parseMessage(data: Data, remaining: Int) -> MessageType? {
+    if data.count < remaining {
```

Comment:
If we don’t handle reading more data than is currently remaining, should we check for `data.count > remaining` and give some indication that allows easier debugging? I’d be fine with a `precondition` (I rarely am 😉) since this shouldn’t be reachable by a user and it’s a serious error in the JSON-RPC connection code that we can’t meaningfully recover from.

Reply:
Precondition it is. FWIW it's not all that difficult to handle having arbitrary reads if we want to (I didn't since it isn't needed, but you could argue it's more resilient).